src/v/cluster/log_eviction_stm.cc

```diff
@@ -40,79 +41,109 @@ log_eviction_stm::log_eviction_stm(
     ssx::spawn_with_gate(
       _gate, [this] { return write_raft_snapshots_in_background(); });
+    ssx::spawn_with_gate(_gate, [this] { return monitor_log_eviction(); });
     return persisted_stm::start();
 }
 
 ss::future<> log_eviction_stm::stop() {
-    _reap_condition.broken();
+    _queue_mutex.broken();
+    _queue.abort(std::make_exception_ptr(ss::abort_requested_exception()));
     co_await persisted_stm::stop();
 }
 
+ss::future<> log_eviction_stm::enqueue_eviction_event(
+  model::offset offset, wait_for_success wait) {
+    co_await _queue_mutex.get_units();
+    co_await _queue.push_eventually(eviction_event{
+      .prefix_truncate_offset = offset, .wait_for_truncation = wait});
+}
+
 ss::future<> log_eviction_stm::write_raft_snapshots_in_background() {
     /// This method is executed as a background fiber and it attempts to write
     /// snapshots as close to effective_start_offset as possible.
     auto gh = _gate.hold();
-    bool wait_with_timeout = false;
+
     while (!_as.abort_requested()) {
         /// This background fiber can be woken-up via apply() when special
         /// batches are processed or by the storage layer when local
         /// eviction is triggered.
         try {
-            static const auto eviction_event_wait_time = 5s;
-            if (wait_with_timeout) {
-                /// If after call to \ref do_write_snapshot() has not
-                /// truncated up until the desired point, it will be OK to
-                /// iterate again after some time to retry. Maybe
-                /// max_collectible_offset has moved forward.
-                co_await _reap_condition.wait(eviction_event_wait_time);
-            } else {
-                /// Last truncation had completed with success, there is no need
-                /// to wait for any other condition then for notify() to be
-                /// called.
-                co_await _reap_condition.wait();
-            }
-        } catch (const ss::condition_variable_timed_out&) {
-            /// There is still more data to truncate
+            co_await _queue.not_empty();
+        } catch (const ss::abort_requested_exception&) {
+            break;
         } catch (const ss::broken_condition_variable&) {
             /// stop() has been called
             break;
+        } catch (const ss::broken_semaphore&) {
+            /// stop() has been called
+            break;
         }
+
+        auto& event = _queue.front();
+        vlog(
+          _log.trace,
+          "processing prefix truncation event with offset: {}, waiting for "
+          "truncate: {}",
+          event.prefix_truncate_offset,
+          event.wait_for_truncation);
+
         auto evict_until = std::max(
-          _delete_records_eviction_offset, _storage_eviction_offset);
+          _delete_records_eviction_offset, event.prefix_truncate_offset);
         auto index_lb = _raft->log().index_lower_bound(evict_until);
         if (!index_lb) {
```

Comment:
I think we need to check the snapshot index before doing early continues. What if a previous event evicted everything, and now index_lb always returns nullopt? I think a delete-records request would block repeatedly get caught in this continue

Reply:
I think if we can safely assume that if the only case this method returns nullopt is if the offset doesn't exist in the log, then we can say the data has already been truncated, and a continue would be ok to do

Reply:
If the current event we're processing is a delete-records event (e.g. immediately after all segments have been GCed by a storage eviction event), wouldn't we get caught in a busy loop here?